library/ssl_msg.c

```diff
@@ -1973,7 +1973,7 @@ int mbedtls_ssl_decrypt_buf(mbedtls_ssl_context const *ssl,
             increment = mbedtls_ct_size_if_else_0(b, increment);
             pad_count += increment;
         }
-        correct = mbedtls_ct_size_if_else_0(mbedtls_ct_uint_eq(pad_count, padlen), padlen);
+        correct = mbedtls_ct_size_if_else_0(mbedtls_ct_uint_eq(pad_count, padlen), correct);
```

Comment:
Why not
```suggestion
        correct &= mbedtls_ct_uint_eq(pad_count, padlen);
```
? (Which is what the code used to look like.) This snippet of code is already pretty complex as it is. The variable `correct` starts out as true, then there are several statements that might change it to false. Using the `&=` operator makes this logic obvious. Using a complicated expression makes it unclear, and made the bug possible in the first place.

I'm not going to request changes here because the top priority is to get this functionally right. But I would request a change if this was new code.

Reply:
Rather than just one change on this line, I think a lot of this needs to be re-written to improve readability. But for the fix I wanted to make the minimum change to make the code correct.

And I'm somewhat annoyed that this didn't cause any tests to fail

Reply:
I think a more proper fix would be to convert `correct` from a `size_t` to a `mbedtls_ct_condition_t` since it is being used as a boolean. The `mbedtls_ct_size_if_else_0()` calls then turn into `mbedtls_ct_bool_and()` which is probably a bit clearer to follow.

I experimented with a quick PR to implement this #8195 . I haven't looked to see if more could/should be done though.